### PR TITLE
fix: display the content of debug explain

### LINKED ISSUE #430: LogicalPlan::inputs function should return the input plan for Explain enum
**Describe the bug**
While trying to get `explain verbose` for a query, I have discovered that the physical plan displays almost nothing. After some investigation, the problem sounds like the [inputs function](https://github.com/apache/arrow-datafusion/blob/14f1eebef068a9e65f556ed74d2b6d98376c97f4/datafusion/src/logical_plan/plan.rs#L318) of `Explain` does not return its logical plan.

Here is the current code:
`LogicalPlan::Explain { .. } => vec![]`

It should have been:
`LogicalPlan::Explain { plan } => vec![plan]`
and the line should be moved above '// plans without inputs'

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by NGA-TRAN**
# Which issue does this PR close?

Closes #430 

 # Rationale for this change

# What changes are included in this PR?
There are 2 main changes:
1. Fix the display of all LogicalPlan's `display_` functions to have them display the plan content for not only executing plans but also the `explain` ones
2. Propagate the optimization down the plan of for the `explain`
 
# Are there any user-facing changes?
. Number 1 won't because they are currently used for debug only
. Number 2 will help other TableProvider to push their predicate down correctly (ref: https://github.com/influxdata/influxdb_iox/issues/1538)

### CODE REVIEW

File: datafusion/tests/sql.rs [hunk 1]
```diff
@@ -1573,6 +1575,101 @@ async fn csv_explain() {
...
+      {\n
+        graph[label=\"LogicalPlan\"]\n
+        2[shape=box label=\"Explain\"]\n
+      }\n
+      subgraph cluster_3\n
+      {\n
+        graph[label=\"Detailed LogicalPlan\"]\n
+        4[shape=box label=\"Explain\\nSchema: [plan_type:Utf8, plan:Utf8]\"]\n
+      }\n
+    }\n
```
Comment (NGA-TRAN) [inline datafusion/tests/sql.rs:1646]: Similarly, graphviz does not have any useful info

File: datafusion/tests/sql.rs [hunk 2]
```diff
@@ -1573,6 +1575,101 @@ async fn csv_explain() {
...
+    // End DataFusion GraphViz Plan";
+    let actual = format!("{}", plan.display_graphviz());
+    assert_eq!(expected.replace("\t","").replace("\n","").replace(" ",""), actual.replace("\t","").replace("\n","").replace(" ",""));
+
+    // Physical plan
+    // Create plan
+    let msg = format!("Creating physical plan for '{}': {:?}", sql, plan);
+    let plan = ctx.create_physical_plan(&plan).expect(&msg);
+    // Verify the text format of the plan
+    let expected = "ExplainExec";
```
Comment (NGA-TRAN) [inline datafusion/tests/sql.rs:1656]: Similarly, display of physical plan has no useful info at all

File: datafusion/tests/sql.rs [hunk 3]
```diff
@@ -1573,6 +1575,101 @@ async fn csv_explain() {
...
+    let actual = format!("{}", displayable(plan.as_ref()).indent());
+    assert_eq!(expected.replace("\t","").replace("\n","").replace(" ",""), actual.replace("\t","").replace("\n","").replace(" ",""));
+
+    // Execute plan
+    let msg = format!("Executing physical plan for '{}': {:?}", sql, plan);
+    let results = collect(plan).await.expect(&msg);
+    // Compare final explain result from execution output
+    let expected = vec![
+        vec!["logical_plan",
+             "Projection: #c1\n  Filter: #c2 Gt Int64(10)\n    TableScan: aggregate_test_100 projection=None"]];
```
Comment (NGA-TRAN) [inline datafusion/tests/sql.rs:1666]: The useful info only available after execute the EXPLAIN. Should everything above include this information?

File: datafusion/tests/sql.rs [hunk 4]
```diff
@@ -1591,6 +1688,105 @@ async fn csv_explain_verbose() {
     assert!(actual.contains("#c2 Gt Int64(10)"), "Actual: '{}'", actual);
 }
 
+#[tokio::test]
+async fn csv_explain_verbose_plans() {
```
Comment (NGA-TRAN) [inline datafusion/tests/sql.rs:1692]: Same behavior for explain verbose

File: datafusion/tests/sql.rs [hunk 5]
```diff
@@ -1573,6 +1575,101 @@ async fn csv_explain() {
...
+    let plan = ctx.create_logical_plan(&sql).expect(&msg);
+    let logical_schema = plan.schema();
+    // 
+    println!("SQL: {}", sql);
+    // Verify schema
+    let expected = "Explain [plan_type:Utf8, plan:Utf8]";
+    let actual = format!("{}", plan.display_indent_schema());
+    assert_eq!(expected, actual);
+    // Verify the text format of the plan
+    let expected = "Explain";
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/tests/sql.rs:1598]: I think including the plan tree like
```
Explain
  Project
    Filter
      Scan
```

Makes more sense than the current output of:
```
Explain
```

File: datafusion/tests/sql.rs [hunk 6]
```diff
@@ -1573,6 +1575,101 @@ async fn csv_explain() {
...
+        graph[label=\"LogicalPlan\"]\n
+        2[shape=box label=\"Explain\"]\n
+      }\n
+      subgraph cluster_3\n
+      {\n
+        graph[label=\"Detailed LogicalPlan\"]\n
+        4[shape=box label=\"Explain\\nSchema: [plan_type:Utf8, plan:Utf8]\"]\n
+      }\n
+    }\n
+    // End DataFusion GraphViz Plan";
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/tests/sql.rs:1615]: I think it should include the actual plan that is being explained

**Review by alamb — COMMENTED** [review]
I think including the input to the `Explain` node makes sense. Any thoughts @Dandandan  or @andygrove ?

File: datafusion/tests/sql.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply datafusion/tests/sql.rs:1598]: Agree. I actually made this happen in a different branch but want to verify with you first. Will bring it in here

**Comment by andygrove** [comment]
> I think including the input to the `Explain` node makes sense. Any thoughts @Dandandan or @andygrove ?

Makes sense to me

File: datafusion/src/logical_plan/plan.rs [hunk 7]
```diff
@@ -438,11 +439,11 @@ impl LogicalPlan {
                 }
                 true
             }
+            LogicalPlan::Explain { plan, .. } => plan.accept(visitor)?,
```
Comment (NGA-TRAN) [inline datafusion/src/logical_plan/plan.rs:442]: This is the fix that enables LogicalPlan's `display_...` functions to show all info of the plan. The newly added tests capture this fix.

File: datafusion/src/logical_plan/plan.rs [hunk 8]
```diff
@@ -329,11 +329,12 @@ impl LogicalPlan {
             LogicalPlan::Limit { input, .. } => vec![input],
             LogicalPlan::Extension { node } => node.inputs(),
             LogicalPlan::Union { inputs, .. } => inputs.iter().collect(),
+            LogicalPlan::Explain { plan, .. } => vec![plan],
```
Comment (NGA-TRAN) [inline datafusion/src/logical_plan/plan.rs:332]: This is the fix that progpagates the optimization down the plan explain

File: datafusion/src/optimizer/filter_push_down.rs [hunk 9]
```diff
@@ -234,6 +234,10 @@ fn split_members<'a>(predicate: &'a Expr, predicates: &mut Vec<&'a Expr>) {
 
 fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
     match plan {
+        LogicalPlan::Explain { .. } => {
+            // push the optimization to the plan of this explain
+            push_down(&state, plan)
+        }
```
Comment (NGA-TRAN) [inline datafusion/src/optimizer/filter_push_down.rs:240]: This is related to the fix to propagate the optimization down

File: datafusion/src/logical_plan/plan.rs [hunk 10]
```diff
@@ -329,11 +329,12 @@ impl LogicalPlan {
...
             // plans without inputs
             LogicalPlan::TableScan { .. }
             | LogicalPlan::EmptyRelation { .. }
-            | LogicalPlan::CreateExternalTable { .. }
-            | LogicalPlan::Explain { .. } => vec![],
+            //| LogicalPlan::Explain { .. }
```
Comment (alamb) [inline datafusion/src/logical_plan/plan.rs:336]:
```suggestion
```

I think we can remove this line entirely (no need to leave it in)

File: datafusion/tests/sql.rs [hunk 11]
```diff
@@ -1573,6 +1575,173 @@ async fn csv_explain() {
...
+    // Both schema has to be the same
+    assert_eq!(logical_schema.as_ref(), optimized_logical_schema.as_ref());
+    //
+    // Verify schema
+    let expected = "Explain [plan_type:Utf8, plan:Utf8]\n
+      Projection: #c1 [c1:Utf8]\n
+          Filter: #c2 Gt Int64(10) [c1:Utf8, c2:Int32]\n
+                TableScan: aggregate_test_100 projection=Some([0, 1]) [c1:Utf8, c2:Int32]";
+    let actual = format!("{}", plan.display_indent_schema());
+    assert_eq!(
```
Comment (alamb) [inline datafusion/tests/sql.rs:1671]: These tests look somewhat challenging to maintain. What do you think about the following pattern (mostly copied from `assert_batches_eq!`:

```rust
let expected_lines = vec![
  "Explain [plan_type:Utf8, plan:Utf8]",
  "  Projection: #c1 [c1:Utf8]",
  ...
 ];

let formatted = plan.display_indent_schema().to_string();
let actual_lines: Vec<&str> = formatted.trim().lines().collect();

assert_eq!(
  expected_lines, actual_lines,
  "\n\nexpected:\n\n{:#?}\nactual:\n\n{:#?}\n\n",
  expected_lines, actual_lines
);
```

**Review by alamb — APPROVED** [review]
The code changes look good to me @NGA-TRAN  -- thank you.

I had a suggestion on how to make the tests perhaps easier to maintain (and easier to see what was different)

File: datafusion/tests/sql.rs (hunk 11 above)
Comment (NGA-TRAN) [reply datafusion/tests/sql.rs:1678]: Addressed

**Review by NGA-TRAN — COMMENTED** [review]
Thanks @alamb for your review comments. I have addressed them.

**Review by alamb — APPROVED** [review]
I think this is looking great @NGA-TRAN  -- thank you

**Review by jorgecarleitao — APPROVED** [review]
LGTM.

Impressive testing, @NGA-TRAN 💯

**Comment by alamb** [comment]
@NGA-TRAN  it looks like this may have failed on windows: https://github.com/apache/arrow-datafusion/pull/434/checks?check_run_id=2720284528

**Comment by NGA-TRAN** [comment]
| @NGA-TRAN it looks like this may have failed on windows: https://github.com/apache/arrow-datafusion/pull/434/checks?check_run_id=2720284528

Silly me. Working on it. Thanks @alamb

**Comment by NGA-TRAN** [comment]
@alamb Finally all checks have passed :)

**Pull request merged**